File: android/app/src/main/java/com/rafaeltmbr/stopwatch/domain/services/impl/LoggingServiceImpl.kt
package com.rafaeltmbr.stopwatch.domain.services.impl

import com.rafaeltmbr.stopwatch.domain.services.LoggingService
import com.rafaeltmbr.stopwatch.domain.services.external_resources.PlatformLoggerFacade


class LoggingServiceImpl(
    private val platformLogger: PlatformLoggerFacade
) : LoggingService {
    override fun debug(tag: String, message: String) = platformLogger.debug(tag, message)

    override fun info(tag: String, message: String) = platformLogger.info(tag, message)

    override fun warn(tag: String, message: String) = platformLogger.warn(tag, message)

    override fun error(tag: String, message: String, exception: Exception?) =
        platformLogger.error(tag, message, exception)
}

File: ios/Stopwatch/domain/services/impl/LoggingServiceImpl.swift
import Foundation

class LoggingServiceImpl: LoggingService {
    func debug(tag: String, message: String) {
        print(formatMessage(severity: "debug", tag: tag, message: message))
    }
    
    func info(tag: String, message: String) {
        print(formatMessage(severity: "info", tag: tag, message: message))
    }
    
    func warn(tag: String, message: String) {
        print(formatMessage(severity: "warn", tag: tag, message: message))
    }
    
    func error(tag: String, message: String, error: (any Error)? = nil) {
        let formatted = formatMessage(severity: "error", tag: tag, message: message)
        if error == nil {
            print(formatted)
            return
        }
        
        print("\(formatted) \(error.debugDescription)")
    }
    
    private func formatMessage(severity: String, tag: String, message: String) -> String {
        let time = ISO8601DateFormatter().string(from: Date())
        return "\(time) \(severity.uppercased()) [\(tag)] \(message)"
    }
}
